File: backends/node/shared/implementations/caching/distributed/redis/src/handlers/r/get.ts
import type Redis from "ioredis";
import { BaseHandler, type IHandlerContext, type RedactionSpec } from "@d2/handler";
import { TK } from "@d2/i18n";
import { D2Result, ErrorCodes, HttpStatusCode } from "@d2/result";
import { redisErrorResult } from "../../redis-error-result.js";
import type { DistributedCache } from "@d2/interfaces";
import { JsonCacheSerializer, type ICacheSerializer } from "../../serialization.js";

export class Get<TValue>
  extends BaseHandler<DistributedCache.GetInput, DistributedCache.GetOutput<TValue>>
  implements DistributedCache.IGetHandler<TValue>
{
  private readonly redis: Redis;
  private readonly serializer: ICacheSerializer<TValue>;

  constructor(redis: Redis, context: IHandlerContext, serializer?: ICacheSerializer<TValue>) {
    super(context);
    this.redis = redis;
    this.serializer = serializer ?? new JsonCacheSerializer<TValue>();
  }

  override get redaction(): RedactionSpec {
    return { outputFields: ["value"] };
  }

  protected async executeAsync(
    input: DistributedCache.GetInput,
  ): Promise<D2Result<DistributedCache.GetOutput<TValue> | undefined>> {
    try {
      const raw = await this.redis.getBuffer(input.key);
      if (raw === null) {
        return D2Result.notFound();
      }
      try {
        const value = this.serializer.deserialize(raw);
        return D2Result.ok({ data: { value } });
      } catch (err: unknown) {
        this.context.logger.warn(
          `Cache deserialization failed for key "${input.key}": ${err instanceof Error ? err.message : String(err)}`,
        );
        return D2Result.fail({
          messages: [TK.common.errors.COULD_NOT_BE_DESERIALIZED],
          statusCode: HttpStatusCode.InternalServerError,
          errorCode: ErrorCodes.COULD_NOT_BE_DESERIALIZED,
        });
      }
    } catch (err: unknown) {
      return redisErrorResult(err);
    }
  }
}

File: backends/dotnet/shared/Implementations/Caching/Distributed/DistributedCache.Redis/Handlers/R/Get.cs
// -----------------------------------------------------------------------
// <copyright file="Get.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.DistributedCache.Redis.Handlers.R;

// ReSharper disable AccessToStaticMemberViaDerivedType
using System.Text.Json;
using D2.Shared.Handler;
using D2.Shared.I18n;
using D2.Shared.Result;
using D2.Shared.Utilities.Serialization;
using Microsoft.Extensions.Logging;
using StackExchange.Redis;
using S = D2.Shared.Interfaces.Caching.Distributed.Handlers.R.IRead;

/// <summary>
/// Handler for retrieving a value from the distributed cache.
/// </summary>
///
/// <typeparam name="TValue">
/// The type of the cached value.
/// </typeparam>
public partial class Get<TValue> : BaseHandler<
        S.IGetHandler<TValue>, S.GetInput, S.GetOutput<TValue>>,
    S.IGetHandler<TValue>
{
    /// <summary>
    /// Cached protobuf parser delegate. Static per closed generic type, so reflection
    /// runs once per TValue regardless of how many <see cref="Get{TValue}"/> instances exist.
    /// </summary>
    private static readonly Lazy<Func<byte[], TValue>> sr_parser = new(BuildParser);

    private readonly IConnectionMultiplexer r_redis;

    /// <summary>
    /// Initializes a new instance of the <see cref="Get{TValue}"/> class.
    /// </summary>
    ///
    /// <param name="redis">
    /// The Redis connection multiplexer.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public Get(
        IConnectionMultiplexer redis,
        IHandlerContext context)
        : base(context)
    {
        r_redis = redis;
    }

    /// <inheritdoc/>
    protected override async ValueTask<D2Result<S.GetOutput<TValue>?>> ExecuteAsync(
        S.GetInput input,
        CancellationToken ct = default)
    {
        try
        {
            // Connect to Redis and get the value.
            var db = r_redis.GetDatabase();
            var redisValue = await db.StringGetAsync(input.Key);

            // If no value was retrieved, return NotFound.
            if (!redisValue.HasValue)
            {
                return D2Result<S.GetOutput<TValue>?>.NotFound();
            }

            // Deserialize the value.
            var value = typeof(TValue).IsAssignableTo(typeof(Google.Protobuf.IMessage))
                ? ParseProtobuf((byte[])redisValue!)
                : JsonSerializer.Deserialize<TValue>(
                    (byte[])redisValue!,
                    SerializerOptions.SR_IgnoreCycles);

            // Return the result.
            return D2Result<S.GetOutput<TValue>?>.Ok(
                new S.GetOutput<TValue>(value));
        }
        catch (RedisException ex)
        {
            LogGetFailed(Context.Logger, ex, input.Key, TraceId);

            return D2Result<S.GetOutput<TValue>?>.ServiceUnavailable(traceId: TraceId);
        }
        catch (JsonException ex)
        {
            LogGetDeserializationFailed(Context.Logger, ex, input.Key, TraceId);

            return D2Result<S.GetOutput<TValue>?>.UnhandledException(
                messages: [TK.Common.Errors.COULD_NOT_BE_DESERIALIZED],
                traceId: TraceId);
        }

        // Let the base handler catch any other exceptions.
    }

    /// <summary>
    /// Logs that a Redis exception occurred while getting a cached value.
    /// </summary>
    [LoggerMessage(EventId = 1, Level = LogLevel.Error, Message = "RedisException occurred while getting value for key '{Key}'. TraceId: {TraceId}")]
    private static partial void LogGetFailed(ILogger logger, Exception ex, string key, string? traceId);

    /// <summary>
    /// Logs that a JSON deserialization exception occurred while getting a cached value.
    /// </summary>
    [LoggerMessage(EventId = 2, Level = LogLevel.Error, Message = "JsonException occurred while deserializing value for key '{Key}'. TraceId: {TraceId}")]
    private static partial void LogGetDeserializationFailed(ILogger logger, Exception ex, string key, string? traceId);

    /// <summary>
    /// Parses a Protobuf message from a byte array using a cached delegate.
    /// </summary>
    ///
    /// <param name="bytes">
    /// The byte array containing the Protobuf message.
    /// </param>
    ///
    /// <returns>
    /// The parsed Protobuf message.
    /// </returns>
    private static TValue ParseProtobuf(byte[] bytes) => sr_parser.Value(bytes);

    /// <summary>
    /// Builds the protobuf parser delegate via reflection (called once per TValue).
    /// </summary>
    private static Func<byte[], TValue> BuildParser()
    {
        var parserProperty = typeof(TValue).GetProperty(
            "Parser",
            System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Static)
            ?? throw new InvalidOperationException(
                $"Type '{typeof(TValue).FullName}' does not have a public static 'Parser' property. Ensure that TValue is a generated protobuf message type.");

        var parser = parserProperty.GetValue(null)
            ?? throw new InvalidOperationException(
                $"The 'Parser' property on type '{typeof(TValue).FullName}' is null. Ensure that TValue is a valid protobuf message type.");

        var parseFromMethod = parser.GetType().GetMethod("ParseFrom", [typeof(byte[])])
            ?? throw new InvalidOperationException(
                $"The 'Parser' object on type '{typeof(TValue).FullName}' does not have a 'ParseFrom(byte[])' method. Ensure that TValue is a valid protobuf message type.");

        return data => (TValue)parseFromMethod.Invoke(parser, [data])!;
    }
}
